Podcast/Authentication.swift

```diff
@@ -0,0 +1,144 @@
+        get {
+            return GIDSignIn.sharedInstance().currentUser?.authentication.accessToken
+        }
+    }
+
+    override init() {
+        super.init()
+        GIDSignIn.sharedInstance().clientID = "724742275706-h8qs46h90squts3dco76p0q6lja2c7nh.apps.googleusercontent.com"
+
+        let profileScope = "https://www.googleapis.com/auth/userinfo.profile"
+        let emailScope = "https://www.googleapis.com/auth/userinfo.email"
+
+        GIDSignIn.sharedInstance().scopes.append(contentsOf: [profileScope, emailScope])
+        facebookLoginManager = LoginManager()
+        GIDSignIn.sharedInstance().delegate = self
+    }
+
+    func signInWithFacebook(viewController: UIViewController, success: (() -> ())? = nil, failure: (() -> ())? = nil) {
+        facebookLoginManager.logIn(readPermissions: [.publicProfile, .email, .userFriends], viewController: viewController) { loginResult in
+            switch loginResult {
+            case .failed(let error):
+                print(error)
+                failure?()
+            case .cancelled:
+                print("User cancelled login.")
+                failure?()
+            case .success(_):
+                success?()
+            }
+        }
+    }
+
+    func signInSilentlyWithGoogle() {
+         GIDSignIn.sharedInstance().signInSilently()
+    }
+    
+    func signInWithGoogle() {
+        GIDSignIn.sharedInstance().signIn()
+    }
+
+    func logout() {
+        GIDSignIn.sharedInstance().signOut()
+        facebookLoginManager.logOut()
+    }
+
+    // Google sign in functionality
+    func handleSignIn(url: URL, options: [UIApplicationOpenURLOptionsKey : Any]) -> Bool {
+        return GIDSignIn.sharedInstance().handle(url,
+                                                 sourceApplication: options[UIApplicationOpenURLOptionsKey.sourceApplication] as? String,
+                                                 annotation: options[UIApplicationOpenURLOptionsKey.annotation])
+    }
+
+    // delegate method for Google sign in - called when sign in is comeplete
+    // awkward to put here but this is how Google requires signing in delegation
+    func sign(_ signIn: GIDSignIn!, didSignInFor user: GIDGoogleUser!, withError error: Error!) {
+        // if we are in the LoginViewController
+        if let window = UIApplication.shared.delegate?.window as? UIWindow, let navigationController = window.rootViewController as? UINavigationController, let viewController = navigationController.viewControllers.first as? LoginViewController {
+            viewController.signInWithGoogle(wasSuccessful: error == nil)
+        } else {
+            if error == nil {
+                // else merge accounts in SettingsPageViewController
+                Authentication.sharedInstance.mergeAccounts(signInTypeToMergeIn: .Google, success: { _,_,_ in
+                    UserSettings.mainSettingsPage.navigationController?.popToRootViewController(animated: false)}
+                , failure: {
```

Comment:
Shouldn't we do something different on failure?

Reply:
I added an error alert